Add Docker Provider type

dPhanekham opened the pull request:
Useful for container to container testing and on-prem to cloud testing

Includes support for

Docker as a provider type
Docker volumes as scratch disks
Local Docker image builds
I will gladly modify or add anything that is needed.

s-deitz commented on perfkitbenchmarker/providers/docker/docker_virtual_machine.py:
```diff
@@ -0,0 +1,303 @@
+    with self.docker_build_lock:
+      image_exists = self._LocalImageExists(self.container_image)
+      if image_exists is False:
+        self._BuildImageLocally()
```
Is the docker image read-only? I'm curious about the case where we might have two Docker virtual machines. Is there any reason to support that?

dPhanekham replied:
A docker image is read-only. This bit of code is to prevent a specific case where Docker image doesn't yet exist on the local machine and you are running a benchmark that uses two containers, then it will attempt to build the Docker image in two threads at the same time, leading to an untagged duplicate image that is never used and just taking up space.

My specific use case for using multiple docker containers was when I was developing/testing a networking benchmark and containers were faster to test on. Other than testing, inter-container communication could be a reasonable thing to benchmark

s-deitz commented on perfkitbenchmarker/providers/docker/docker_virtual_machine.py:
```diff
@@ -0,0 +1,303 @@
+
+  @vm_util.Retry(poll_interval=10, max_retries=10)
+  def _Exists(self):
+    """
```
There is a bit of inconsistency in the docstring style. See https://www.python.org/dev/peps/pep-0257/ for formatting of single line and multiline docstrings and when to use either. In particular, single line docstring comments should be on the same line as the """ delimiters.

A single line here might be:
"""Returns whether the container is up and running."""

s-deitz commented on perfkitbenchmarker/providers/docker/docker_virtual_machine.py:
```diff
@@ -0,0 +1,303 @@
+    return info, returnCode
+
+  def _ConfigureProxy(self):
+    """
```
Single line summary and then elaborate.

https://www.python.org/dev/peps/pep-0257/

s-deitz commented on perfkitbenchmarker/providers/docker/docker_virtual_machine.py:
```diff
@@ -0,0 +1,303 @@
+      sysctl_params: dict - the keys and values to write
+    """
+    logging.warn("Sysctl flags NOT applied to Docker container. "
+                 "Please use --docker_sysctl_flags if sysctl need to be applied")
```
Ah, this is the override I was thinking of above. Can you use the same flag and disable applying it here?

dPhanekham replied:
switched to use --sysctl flag. removed --docker_sysctl_flags

ehankland commented on perfkitbenchmarker/providers/docker/docker_virtual_machine.py:
```diff
@@ -0,0 +1,303 @@
+    Build Container Image Locally
+    """
+
+    # Dockerfiles located at PerfKitBenchmarker/docker/<containerImage>/Dockerfile
```
we already have perfkitbenchmarker/data/docker/<containerImage>/ - probably don't want two directories so either move the existing image/related functions to the new directory or switch to the /data/ directory

dPhanekham replied:
moved Dockerfile into /perfkitbenchmarker/data/docker/

ehankland commented on perfkitbenchmarker/providers/docker/docker_virtual_machine.py:
```diff
@@ -0,0 +1,303 @@
+    # Dockerfiles located at PerfKitBenchmarker/docker/<containerImage>/Dockerfile
+    directory = os.path.dirname(
+        data.ResourcePath(os.path.join('docker', self.container_image, 'Dockerfile')))
+    build_cmd = [
```
Would be nice if these build functions could be shared with https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/master/perfkitbenchmarker/container_service.py#L321

dPhanekham replied:
I agree that using those build functions would be nice, and I looked into it when I first wrote this. But these are part of BaseContainerRegistry, and the Docker provider type doesn't really have a registry since the image it uses is just built locally.

dPhanekham commented on perfkitbenchmarker/providers/docker/docker_virtual_machine.py:
```diff
@@ -0,0 +1,303 @@
+DEFAULT_DOCKER_IMAGE = 'pkb/ubuntu16_ssh'
+
+
+class DockerVirtualMachine(virtual_machine.BaseVirtualMachine):
```
I think this is a good idea. I've switched to a naming convention more like the Mesos provider.

dPhanekham commented on perfkitbenchmarker/providers/docker/docker_virtual_machine.py:
```diff
@@ -0,0 +1,303 @@
+    if FLAGS.docker_custom_image:
+      self.container_image = FLAGS.docker_custom_image
+    if FLAGS.docker_sysctl_flags:
+      self.docker_sysctl_flags = FLAGS.docker_sysctl_flags
```
Switched to use --sysctl flag already present, updated dependent functions accordingly

dPhanekham commented on docker/pkb/ubuntu16_ssh/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+#Dockerfile to use with Docker provider
```
moved to perfkitbenchmarker/data/docker

dPhanekham commented on perfkitbenchmarker/providers/docker/docker_resource_spec.py:
```diff
@@ -0,0 +1,53 @@
+    result = super(DockerSpec, cls)._GetOptionDecoderConstructions()
+    result.update({
+        'docker_provider_cpus': (option_decoders.FloatDecoder, {'default': 0}),
+        'docker_provider_memory_mb': (option_decoders.IntDecoder, {'default': 0}),
```
I have removed these flags in favor custom_virtual_machine_spec stuff. Can now use --machine_type flag to supply custom cpu/memory.

dPhanekham commented:
Thank you for all of your feedback. It has been very useful for improving my code. I think I have now addressed all of your comments. For the simpler comments I just made the changes, and for others I both made changes and responded to the comment directly. Let me know if there are any further changes that need to be made.

s-deitz commented on perfkitbenchmarker/providers/docker/docker_container_instance.py:
```diff
@@ -0,0 +1,319 @@
+    DebianBasedDockerContainer, linux_virtual_machine.Ubuntu1604Mixin):
+  DEFAULT_IMAGE = UBUNTU_IMAGE
+
+# class Ubuntu1404BasedDockerVirtualMachine(
```
Suggest dropping this commented out code and perhaps adding a note about what it will take to add support for Ubuntu 14 and Ubuntu 18.

Drop Ubuntu 17.10 since it is not a long-term support version and is no longer supported.

s-deitz commented:
There are conflicts in this PR. If you are able to resolve those conflicts, we can merge this change.

dPhanekham commented:
I've done a rebase with master. I think it should work now.

s-deitz commented:
Thanks. I have merged these changes.
